### PR TITLE
feat: support 2d nonces in EVM

### PR DESCRIPTION
**Pull request by klkvr**
Changes in the PR:
- `nonceKey` is changed to `U256`
- AA transactions with non-zero `nonceKey` are now rejected at the txpool level instead of returning an error during decoding.
- Logic for bumping 2D nonce is added to the EVM.
- AA transactions nonce is now always bumped before execution to fix a bug when reverting CREATE transaction did not bump a nonce. A test is added for this edge case

### CODE REVIEW

File: crates/transaction-pool/src/validator.rs [hunk 1]
```diff
@@ -31,13 +32,25 @@ where
...
+        // Reject system transactions, those are never allowed in the pool.
         if transaction.inner().is_system_tx() {
             return TransactionValidationOutcome::Error(
                 *transaction.hash(),
                 InvalidTransactionError::TxTypeNotSupported.into(),
             );
         }
 
+        // Reject AA transactions with non-zero nonce keys, we don't have
+        // a nice way to track them in the pool yet.
```
Comment (mattsse) [inline crates/transaction-pool/src/validator.rs:44]: mental note here, this should be removed once we have #569 ?

File: crates/revm/src/handler.rs [hunk 2]
```diff
@@ -371,19 +372,78 @@ where
...
+                .increment_nonce(&tx.caller(), nonce_key)
+                .map_err(|err| match err {
+                    TempoPrecompileError::Fatal(err) => EVMError::Custom(err),
+                    err => TempoInvalidTransaction::NonceManagerError(err.to_string()).into(),
+                })?;
+        } else {
+            // Bump the nonce for calls. Nonce for CREATE will be bumped in `make_create_frame`.
+            //
+            // Always bump nonce for AA transactions.
+            if tx.aa_tx_env.is_some() || tx.kind().is_call() {
```
Comment (legion2002) [inline crates/revm/src/handler.rs:561]: Why do we always bump nonce for AA tx type here?
More accurate would be to bump if all calls in the batch, are of type call

File: crates/revm/src/handler.rs [hunk 3]
```diff
@@ -371,19 +372,78 @@ where
...
+                            },
+                        )
+                        .into());
+                    }
+                    _ => {}
+                }
+            }
+
+            // Always increment nonce for AA transactions with non-zero nonce keys.
+            nonce_manager
```
Comment (legion2002) [inline crates/revm/src/handler.rs:551]: should we also disallow any CREATE calls in the batch here?

File: crates/revm/src/handler.rs (hunk 2 above)
Comment (klkvr) [reply crates/revm/src/handler.rs:561]: we can't rely on CREATE item to bump the nonce because right now if that CREATE reverts we don't retain the nonce bump in the journal

this is the case on main right now and I don't see a nice way to fix this

File: crates/revm/src/handler.rs (hunk 3 above)
Comment (klkvr) [reply crates/revm/src/handler.rs:551]: why would we want to disallow them specifically for non-zero keys?

File: crates/revm/src/handler.rs (hunk 2 above)
Comment (legion2002) [reply crates/revm/src/handler.rs:561]: but doesn't this mean that for successful CREATE, the nonce is bumped twice? 
So if an AA tx has just a single CREATE, then nonce is bumped by 2.

But ig it doesn't that matter, it is the same behavior as an EVM CREATE.

Comment (klkvr) [reply crates/revm/src/handler.rs:561]: > So if an AA tx has just a single CREATE, then nonce is bumped by 2.

yep

> But ig it doesn't that matter, it is the same behavior as an EVM CREATE.

it is not, if you send a eip1559 tx that does a CREATE, nonce will only be numped once

Comment (legion2002) [reply crates/revm/src/handler.rs:561]: yeah by EVM create I meant, if you have a delegated account, and that account uses the CREATE opcode.

Comment (klkvr) [reply crates/revm/src/handler.rs:561]: I guess we could special-case this — i.e if AA tx's first frame is CREATE and it is using protocol nonce we will rely on CREATE frame to bump it, and then if it fails we will still forcefully bump it

though this is a bit inconsistent, because for non-zero keys we'd still bump nonce 2 times — once for the key and once for the protocol nonce

File: crates/revm/src/handler.rs (hunk 3 above)
Comment (legion2002) [reply crates/revm/src/handler.rs:551]: Oh yeah nvm

File: crates/revm/src/handler.rs (hunk 2 above)
Comment (legion2002) [reply crates/revm/src/handler.rs:561]: Yeah I think it's fine.
Don't think any major applications assume the sequentiality of protocol nonce, if someone wants complete control over the nonce, they would just use CREATE2
